File: leetcode/questions/347/java/347.java
import java.util.*;

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> count = new HashMap<>();
        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> Integer.compare(a[1], b[1]));

        for (int num: nums) {
            count.put(num, count.getOrDefault(num, 0) + 1);
        }

        count.forEach((key, val) -> {
            heap.offer(new int[] {key, val});
            if(heap.size() > k) {
                heap.poll();
            }
        });

        int heapSize = heap.size();
        int[] ans = new int[heap.size()];

        for (int i = 0; i < heapSize; ++i) {
            int[] pair = heap.poll();
            ans[i] = pair[0];
        }

        return ans;
    }
}

File: leetcode/questions/347/python/347.py
from typing import List
from collections import Counter

import heapq

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        heap = []
        count = Counter(nums)

        for num, freq in count.items():
            heapq.heappush(heap, (freq, num))
            if len(heap) > k:
                heapq.heappop(heap)

        return [num for _, num in heap]
